### ISSUE TITLE
Implement Polyphony for piano

### ISSUE THREAD

**Issue by abhinav3967** [QuadCubedStudios/GoDAW#35]
# Polyphony

The piano instrument as of now can only play one note at a time. It also stops all the music that is playing if stop_note is called on the piano instrument. Playing multiple notes at the same time is important

**Comment by Technohacker** [comment]
This may have to be instrument-specific, since certain instruments may be monophonic by design, or have a limited number of voices (pure chiptune)

**Comment by abhinav3967** [comment]
polyphony can be a variable in instrument. the value of polyphony will be the no. of notes that can be played at a time

**Comment by razetime** [comment]
if everything is polyphonic, then theres always the option of playing one note at a time. why have an unnecessary extra option?

**Comment by Technohacker** [comment]
I've added a method to the Instrument class to let the instrument take an AudioStreamPlayer, with defaults suited for GoDAW. This'll probably be enough, with the instrument deciding how to use those players to play notes

**Comment by abhinav3967** [comment]
> if everything is polyphonic, then theres always the option of playing one note at a time. why have an unnecessary extra option?

because we want to cancel out a note if two notes are pressed. completely optional tho.

**Comment by abhinav3967** [comment]
> I've added a method to the Instrument class to let the instrument take an AudioStreamPlayer, with defaults suited for GoDAW. This'll probably be enough, with the instrument deciding how to use those players to play notes

so polyphony has to be implemented by the instrument maker?

**Comment by Technohacker** [comment]
> > I've added a method to the Instrument class to let the instrument take an AudioStreamPlayer, with defaults suited for GoDAW. This'll probably be enough, with the instrument deciding how to use those players to play notes
> 
> so polyphony has to be implemented by the instrument maker?

Yup, and commit 3680c96 shows one example of this with TripleOsc

**Comment by abhinav3967** [comment]
Nice, Now can this issue be closed?

**Comment by Technohacker** [comment]
Not yet, I may need a convenience method for instrument-makers for later :)

**Comment by abhinav3967** [comment]
oh! okay

**Comment by IsotoxalDev** [comment]
So any update on this?
